Revision of Section 6 to reflect Things and Consumers

mkovatsc opened the pull request:
This is work in progress.

Section 6.1 already contains new text to define Things and Consumers independent from client/server roles.

The change also includes regrouping to reflect clarifications around Interaction Affordances, Hypermedia Controls, and Protocol Bindings as discussed during the joint IETF workshop.

The section for system components introducing Servients is now 6.7 (for @takuki and @mryuichi )

mkovatsc commented:
[Preview PR Changes](https://cdn.staticaly.com/gh/w3c/wot-architecture/mkovatsc-revision-section-6/index.html?env=dev)

mkovatsc commented:
Please already start to review the updates to the terminology. Besides Consumer and Removing WoT Client/Server, they should be minor tweaks or additions. This was partly necessary to improve the relation between the definitions.

mlagally commented on index.html:
```diff
@@ -496,83 +504,62 @@ <h1>Terminology</h1>
-                <dfn data-lt="WoT Thing Description">Thing
-                    Description</dfn>
-            </dt>
-            <dd>Structured data describing a Thing. A TD includes
```
Perhaps the current definition is not ideal and should be improved, however I think we should keep "Thing Description" in the Terminology section.

mkovatsc commented:
Completed update of Section 6, except the new 6.7 subsection.

mlagally commented on index.html:
```diff
@@ -1512,113 +1499,146 @@ <h1>WoT Architecture</h1>
+                The format can be processed either through classic JSON libraries or a JSON-LD processor, as the underlying information model is graph-based and its serialization compatible with JSON-LD 1.1 [[?json-ld-syntax]].
+                A <a>TD</a> is instance-specific (i.e., describes an individual Thing, not types of Things)
+                and is the default Web representation of a <a>Thing</a>.
+                There MAY be other Web representations of a <a>Thing</a> such as an HTML-based user interface or simply an image of the physical entity.
```
Please consider to generalize the term "Web representation" - it is not Web-specific and could be used for closed (non-Web) scenarios as well.
You could call it external representation.

mkovatsc replied:
On the other hand, such representations with Media Type are exactly one of those "Web technologies" which found application in the other systems you are hinting at.

mkovatsc replied:
Now added or even non-Web representations in closed systems

mlagally commented on index.html:
```diff
@@ -1512,113 +1499,146 @@ <h1>WoT Architecture</h1>
+                Simple figure to illustrate Consumer--Thing interaction.
+            </p>
+            <p>
+                A <a>Thing</a> can also be the abstraction of a virtual entity,
```
Is a thing an abstraction of a virtual entity or is the virtual entity already the abstraction?

mkovatsc replied:
The virtual entity is the composition. The composition can be treated under the same Thing abstraction as physical entities.

mkovatsc replied:
Broke apart sentence to be clear.

mlagally commented on index.html:
```diff
@@ -1512,113 +1499,146 @@ <h1>WoT Architecture</h1>
+                Linking does not only apply to hierarchical Things, but relations between Things and other resources in general.
+                Link relation types express how Things relate, for instance, a switch controlling a light or a room monitored by a motion sensor.
+                Other resources related to a Thing can be manuals, catalogs for spare parts, CAD files, a graphical UI, or any other document on the Web.
+                Overall, Web linking among Things makes the Web of Things browsable.
```
browsable -> navigable ?

mkovatsc replied:
It's the **Web** of Things. Why dance to hide this? For me "browsable" also captures the usability we are used to from the Web.

mkovatsc replied:
Clarified that machines can also browse

mlagally commented on index.html:
```diff
@@ -1512,113 +1499,146 @@ <h1>WoT Architecture</h1>
+            </p>
+            <p>
+                Things must be hosted on networked system components with a software stack to realize interaction through a network-facing interface, the <a>WoT Interface</a> of a <a>Thing</a>.
+                One example of this is an HTTP server running on an embedded device with sensors and actutors interfacing the physical entity behind the Thing abstraction.
```
actuator

mkovatsc replied:
Commit incoming

mlagally commented on index.html:
```diff
@@ -1512,113 +1499,146 @@ <h1>WoT Architecture</h1>
+                but which points to the <a>WoT Interface</a> provided by the <a>Intermediary</a>.
+                <a>Intermediaries</a> may also augment existing <a>Things</a> with additional capabilities or compose a new <a>Thing</a> out of multiple available <a>Things</a>,
+                thereby forming a virtual entity.
+                To <a>Consumers</a>, <a>Intermediaries</a> look like <a>Things</a>, as they possess <a>WoT Thing Descriptions</a> and provide a <a>WoT Interface</a>,
```
perhaps "look and behave like things", but i would argue that they are just things with the additional virtualisation aspect.

mkovatsc replied:
Then what is this virtualization aspect? How would you define it?

mlagally commented on index.html:
```diff
@@ -1512,113 +1499,146 @@ <h1>WoT Architecture</h1>
+                thereby forming a virtual entity.
+                To <a>Consumers</a>, <a>Intermediaries</a> look like <a>Things</a>, as they possess <a>WoT Thing Descriptions</a> and provide a <a>WoT Interface</a>,
+                and hence might be indistinguishable from <a>Things</a> in a layered system architecture like the Web [[?REST]].
+                An identifier in the <a>WoT Thing Description</a> MUST allow for the correlation of multiple <a>TDs</a> representing the same original <a>Thing</a> or ultimately unique physical entity.
```
physical or virtual entity. Intermediaries also have TDs with a unique id.

mkovatsc replied:
The may have a unique **Thing ID** if the choose to break the connection to the original Thing. There are use cases where Intermediaries want to preserve it and keep the same ID.

It is not a TD ID.

mlagally commented on index.html:
```diff
@@ -1512,113 +1499,146 @@ <h1>WoT Architecture</h1>
-                mapping of those interactions to concrete protocols
-                (what we call the Protocol Binding), and its security
+                A Web Thing has four architectural
+                aspects of interest: its behavior, its <a>Interaction Affordances</a>, and its security
```
Please add Protocol Bindings to add the 4th aspect to this sentence

mkovatsc replied:
This is not from me and I am waiting for @mmccool in #220

mlagally commented on index.html:
```diff
@@ -1642,192 +1662,172 @@ <h2>Architectural Aspects of a Web Thing</h2>
+                can be modeled.
+            </p>
+            <p>
+                In addition to navigation affordances (i.e., Web links),
```
The term "navigation affordance" has not been explicitly introduced so far.
We need another sentence that explains what a "navigation affordance" is.

Btw.: It would be good to have an example of two linked things in the TD spec.

mkovatsc replied:
See text above under Links and second TODO figure

mlagally commented on index.html:
```diff
@@ -1642,192 +1662,172 @@ <h2>Architectural Aspects of a Web Thing</h2>
+                <p>
+                    Actions MAY contain up to two data schemas,
+                    one for optional input parameters and one for output results,
+                    if the data is not fully specified by the Protocol Binding used (e.g., through a Media Type).
```
data schema or data format?

mkovatsc replied:
Data schema if no Media Type (or specialization such as Content Type) given

mlagally commented on index.html:
```diff
@@ -1642,192 +1662,172 @@ <h2>Architectural Aspects of a Web Thing</h2>
+                    Examples of Actions are changing multiple Properties simultaneously,
+                    changing Properties over time such as fading the brightness of a light (dimming)
+                    or with a process that shall not be disclosed such as a proprietary control loop algorithm,
+                    or invoking a long-lasting process such as priting a document.</p>
```
typo

mkovatsc replied:
Commit incoming

mlagally commented on index.html:
```diff
@@ -1642,192 +1662,172 @@ <h2>Architectural Aspects of a Web Thing</h2>
+                </p>
+                <p>
+                    Link relation types are either IANA-registered tokens adhering to the ABNF
+                    <code style="white-space: nowrap;">LOALPHA *( LOALPHA / DIGIT / "." / "-" )</code>
```
it would be useful to have examples and link to the  IANA registry for these types.

mkovatsc replied:
Added

mlagally commented:
@mkovatsc Impressive work. I added several comments in-line, please take a look.

mkovatsc commented on index.html:
```diff
@@ -1642,192 +1662,172 @@ <h2>Architectural Aspects of a Web Thing</h2>
+            <p>
+                On the Web, an affordance is the simultaneous presentation of information and controls,
+                such that the information becomes the affordance through which the user obtains choices.
+                For humans, the information is usally text or images describing or decorating a hyperlink.
```
usually
Was hard to find... :)

mkovatsc commented:
I addressed most of the review comments. There are 3 still unresolved:
* "Intermediaries are just things with the additional virtualisation aspect"
* Thing ID, not TD ID
* Missing feedback by @mmccool for #220
